src/DynamoCore/Graph/Nodes/NodeModelExtensions.cs

```diff
@@ -23,6 +23,16 @@ internal static void VisibleUpstreamNodes(this NodeModel node, List<NodeModel> g
             }
         }
 
+        /// <summary>
+        /// Provide the upstream nodes that are imediately connected.
+        /// </summary>
+        /// <param name="node"></param>
+        /// <returns></returns>
+        internal static HashSet<NodeModel> ImediateUpstreamNodes(this NodeModel node)
+        {
+            return node.InPorts.SelectMany(p => p.Connectors.Select(c => c.Start.Owner)).ToHashSet();
```

Comment:
by default does this use reference equality?

Reply:
I am not sure

Reply:
I guess it doesn't matter - it's unlikely nodes with the same id would be in the graph twice...